adds getting started cookbook for `ion-schema-rust` (#80)
amzn/ion-schema: docs/cookbook/ion-schema-rust-getting-started.md

    * [How to create a `SchemsSystem` with given authority or list of authorities?](#how-to-create-a-schemasystem-with-given-authority-or-list-of-authorities)
* [How to load a `Schema`?](#how-to-load-a-schema)
    * [Loading a schema](#loading-a-schema)
* [How to create a schema programmatically(`IslSchema`)?](#how-to-create-a-schema-programmaticallyislschema)
    * [How to create an `IslType`?](#how-to-definecreate-an-isltype)
    * [How to create `IslSchema`?](#how-to-create-islschema)
* [How to validate an Ion value using a `Schema`?](#how-to-validate-an-ion-value-using-a-schema)

## How to use this cookbook?
This cookbook explains how to create all the necessary structures required to perform Ion schema validation. It gives sample code with each section explaining the usage of that operation or structure.
In general, the process of Ion schema validation contains two steps. 
1. The first step is to load a correct Ion schema. In this step, we use a `SchemaSystem` which verifies the syntax of given schema file for its correctness.
2. The second step is to use this generated schema from #1 and validate Ion values using the type definitions defined within the schema.

> This cookbook starts from examples for #1(including examples for all the required structures to be created to complete #1) and at the end adds examples for #2(How to validate Ion value using `ion-schema-rust`)

Reviewer: ```suggestion
This cookbook starts from examples for #1 (including examples for all the required structures to be created to complete #1) and at the end adds examples for #2 (How to validate Ion value using `ion-schema-rust`)
```

* **Schema:** Schema is a collection of 0 or more Types. It contains an optional header, type definitions/ types and an optional footer. The schema header can optionally contain imports for that schema.
* **Import:** An Import allows types from other schemas to be used within a schema definition. These imports are usually defined in the schema header.

## How to create a `SchemaSystem`?

`ion-schema-rust` requires you to create an `SchemaSystem` in order to load a schema from given file location. 
Hence, Creating a `SchemaSystem` is the first step to loading a schema file for validation. 

### Creating `DocumentAuthority`:

In general, users will create a `DocumentAuthority` and then use it to build the `SchemaSystem`. Then this `SchemaSystem` is used instantiating instances of `Schema`.
There are two types of `DocumentAuthority` available:

* `FileSystemDocumentAuthority` 
    * This authority allows to specify the base path to where all the schema files are saved (e.g. `/home/USER/schemas/`)

>     * Adding this authority your `SchemaSystem` would allow to resolve all schema ids (relative path to schema file, e.g. `my_schema.isl`) that are within this authority’s base path(e.g. `/home/USER/schemas/`)

Reviewer: ```suggestion
    * Adding this authority your `SchemaSystem` would allow to resolve all schema ids (relative path to schema file, e.g. `my_schema.isl`) that are within this authority’s base path (e.g. `/home/USER/schemas/`)
```

>     * In order to load a schema id(e.g. `my_schema.isl`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema.isl")`

Reviewer: ```suggestion
    * In order to load a schema id (e.g. `my_schema.isl`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema.isl")`
```

* `MapDocumentAuthority`
    * This authority allows to specify a `HashMap` of schema ids as a key and the schema as value. 
    * Adding this authority your `SchemaSystem` would allow to resolve all schema ids that are within this authority’s map keys.

>     * In order to load a schema id(e.g. `my_schema`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema")`

Reviewer: ```suggestion
    * In order to load a schema id (e.g. `my_schema`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema")`
```

Programmatic construction of Ion schema refers to an internal model representation of Ion schema and it mostly resembles to the [grammar](https://amzn.github.io/ion-schema/docs/spec.html#grammar) specified in Ion schema spec.

### How to define/create an `IslType`?
For an Ion schema type definition like:
```
type:: {
     name:my_type_name,
     type: int,
     all_of: [
         { type: bool }
     ]
}
```
The `IslType` can be created as shown below:
```rust
let isl_type = IslType::named(
    // represents the `name` of the defined type
    "my_type_name".to_owned(),
    vec![
        // represents the `type: int` constraint
        IslConstraint::type_constraint(
            IslTypeRef::named("int")
        ),
        // represents `all_of` with anonymous type `{ type: bool }` constraint
        IslConstraint::all_of(
            vec![
                IslTypeRef::anonymous(
                    vec![
                        IslConstraint::type_constraint(
                            IslTypeRef::named("bool")
                        )
                    ]
                )
            ]
        )
    ]
);
```

### How to create `IslSchema`?

```rust

> // The `isl-type` defiend in the previous section cna be used here

Reviewer: ```suggestion
// The `isl-type` defined in the previous section can be used here
```

let isl_schema = IslSchema::new(vec![], vec![isl_type], vec![]);
```

## How to validate an Ion value using a `Schema`?

_Note: The schema file used in this section is previously defined in `How to load a schema?` section_

> ### How to validate an Ion value using `Schema`?

Reviewer: This is redundant.
```suggestion
```

// Retrieve a particular type from this schema
let type_ref: TypeRef = schema.get_type("my_int_type").unwrap();

let valid_element: OwnedElement = 5.into();
let invalid_element: OwnedElement = 5e3.into();
let invalid_document_element: Vec<OwnedElement> = vec![5.into(), true.into(), 6e3.into()];

// Validate data based on the type: 'my_int_type'
check_value(&valid_element, &type_ref); // this validation passes as the value satisfies integer type constraint
check_value(&invalid_element, &type_ref); // this returns violation as 'my_int_type' expects an integer value
check_value(&invalid_document_element, &type_ref); // this returns violation as 'my_int_type' expects an integer value

// Verify if the given value is valid and print violation for invalid value
fn check_value<I: Into<IonSchemaElement> + Debug + Clone>(value: I, type_ref: &TypeRef) {
    let validation_result: ValidationResult = type_ref.validate(value.to_owned());
    if let Err(violation) = validation_result {
        println!("{}", value.into());
        println!("{:#?}", violation);
    }
}
```

### Output
When run, the code above produces the following output:

> ```rust

Reviewer: Technically, this is output, not Rust.
````suggestion
```
````

---
title:  Getting started with `ion-schema-rust`
---
# {{ page.title }}
_(Applies to Ion Schema 1.0.)_

> This is a getting started cookbook for `ion-schema-rust`. It includes all the examples for using `ion-schema-rust` like loading a schema, using Ion schema ot validate Ion values etc.

Reviewer: ```suggestion
This is a getting started cookbook for `ion-schema-rust`. It includes all the examples for using `ion-schema-rust` like loading a schema, using Ion schema to validate Ion values, etc.
```

## Terms

* **Schema Id**: A unique identifier provided to each schema.
* **Authority:** Authority is responsible for resolving particular class of schema identifiers. One example of authority is file system authority which resolves schema ids to files relative to a base path.
* **Schema System:**  Provides methods for instantiating instances of Schema. Requests each of the provided Authorities, in order to resolve the requested schema id until one successfully resolves it. 
* **Schema:** Schema is a collection of 0 or more Types. It contains an optional header, type definitions/ types and an optional footer. The schema header can optionally contain imports for that schema.
* **Import:** An Import allows types from other schemas to be used within a schema definition. These imports are usually defined in the schema header.

## How to create a `SchemaSystem`?

`ion-schema-rust` requires you to create an [`SchemaSystem`](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html) in order to load a schema from given file location. 
Hence, Creating a `SchemaSystem` is the first step to loading a schema file for validation. 

### Creating `DocumentAuthority`:

> In general, users will create a [`DocumentAuhtority`](https://docs.rs/ion-schema/0.4.0/ion_schema/authority/trait.DocumentAuthority.html) and then use it to build the [SchemaSystem](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html). Then this [SchemaSystem](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html) is used instantiating instances of [Schema](https://docs.rs/ion-schema/0.4.0/ion_schema/schema/struct.Schema.html).

Reviewer: ```suggestion
In general, users will create a [`DocumentAuthority`](https://docs.rs/ion-schema/0.4.0/ion_schema/authority/trait.DocumentAuthority.html) and then use it to build the [SchemaSystem](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html). Then this [SchemaSystem](https://docs.rs/ion-schema/0.4.0/ion_schema/system/struct.SchemaSystem.html) is used instantiating instances of [Schema](https://docs.rs/ion-schema/0.4.0/ion_schema/schema/struct.Schema.html).
```

Also, should these links be git blob permalinks instead of links to a particular tag? (I'm not entirely sure myself.)

I actually think that sample code would be more useful than links to the code. The links to the code can quickly go out of date whenever a new version is released, whereas sample code only needs to change as the actual APIs change.

Author: I wanted to link the terms used here to rust crate docs. Although it does mean that these links could go out of date. I can remove these links to rust docs for the purpose of this document as this document already has a `Terms` section where all the definitions are defined.

    * This authority allows to specify the base path to where all the schema files are saved (e.g. `/home/USER/schemas/`)
    * Adding this authority your `SchemaSystem` would allow to resolve all schema ids (relative path to schema file, e.g. `my_schema.isl`) that are within this authority’s base path(e.g. `/home/USER/schemas/`)
    * In order to load a schema id(e.g. `my_schema.isl`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema.isl")`
* [`MapDocumentAuthority`](https://docs.rs/ion-schema/0.4.0/ion_schema/authority/struct.MapDocumentAuthority.html)
    * This authority allows to specify a `HashMap` of schema ids as a key and the schema as value. 
    * Adding this authority your `SchemaSystem` would allow to resolve all schema ids that are within this authority’s map keys.
    * In order to load a schema id(e.g. `my_schema`) with this `FileSystemDocumentAuthority` that is added to your `SchemaSystem` would be as simple as: `schema_system.load_schema("my_schema")`

_Note: A single `SchemaSystem` can contain multiple `DocumentAuthority`s_

```rust

>   // An example of `MapDocumentAuhtority` with `(schema_id, schema content)` pair

Reviewer: ```suggestion
  // An example of `MapDocumentAuthority` with `(schema_id, schema content)` pair
```

   (
        "my_schema", <--- schema id
        r#"                     _
            schema_header::{     |
            }                    |
            type::{              |  
                name: my_int,    |--> Ion schema as raw string
                type: int,       |
            }                    |
            schema_footer::{     |
            }                    |
        "#,                     -
  ),
```

### How to create `FileSystemDocumentAuthority`?
Creating a `FileSystemAuthority` requires to pass a base path where all the schema files resides.
```rust
// In this example, it is assumed that all the schema files that will
// later be used to load a schema are inside `sample_schemas` folder

let file_system_document_authority = FileSystemDocumentAuthority::new(Path::new(
    "sample_schemas",
));
```

### How to create `MapDocumentAuthority`?
Creating a `MapDocumentAuthority` requires to pass a `HashMap` with key-value pair of `(schema_id, schema content)`.

> ```rust

Reviewer: Actually, I think this example makes the example on L54-L69 redundant. I think you should keep this one and get rid of the other one.

                imports: [{ id: "sample_decimal.isl", type: my_decimal, as: other_decimal }],
            }
            type::{
                name: my_int,
                type: int,
            }
            type::{
                name: my_number,
                all_of: [
                    my_int,
                    other_decimal,
                 ],
            }
            schema_footer::{
           }
        "#,
  ),
  (
        "sample_decimal.isl",
        r#"
            schema_header::{
                imports: [],
            }
            type::{
                name: my_decimal,
                type: decimal,
             }
             schema_footer::{
             }
        "#,
   ),
];

// Create a MapDocumentAuthority using a map like above with 
// schema id as key and schema as value
let map_document_authority = MapDocumentAuthority::new(map_authority);
```

Finally, next code block shows how to create a `SchemSystem` using `FileSystemDocumentAuthority`

```rust
// Create authorities vector containing all the authorities that will be used to load a schema based on schema id
let document_authorities: Vec<Box<dyn DocumentAuthority>> = vec![Box::new(

>     FileSystemDocumentAuthority::new(Path::new("sample_schemas")), <--- provide a path to the authority base folder containing schemas

Reviewer: Makes your note into a valid Rust comment:
```suggestion
    FileSystemDocumentAuthority::new(Path::new("sample_schemas")), // <--- provide a path to the authority base folder containing schemas
```

)];

// Create a new schema system using given document authorities
let mut schema_system = SchemaSystem::new(document_authorities);
```

> ## `DocuementAuthority` operations on a `SchemaSystem`

Reviewer: ```suggestion
## `DocumentAuthority` operations on a `SchemaSystem`
```

### How to get all the authorities that are there for a given `SchemaSystem`?

> 

Reviewer: Would be worth mentioning _why_ you might want to do this.

In order to get all the authorities in a `SchemaSystem`:

```rust
// assuming the SchemSystem in built into variable: `schema_system`
let result = schema_system.authorities()
```

Output of the above operation will have following vector saved inside `result`:
```rust
vec![Box::new(
    FileSystemDocumentAuthority::new(Path::new("sample_schemas")),
)]
```

### How to add new authority into a `SchemaSystem`?

> 

Reviewer: Would be worth adding an explanation of _why_ you might want to do this, and where that `authority` goes in terms of priority when resolving SchemaIDs.

Reviewer: This is what I mean by "where that authority goes in terms of priority when resolving SchemaIDs":
```suggestion
When attempting to resolve a schemaId, `Authority`s are checked in the order in which they were added to the `SchemaSystem`, so any new `Authority` is added with a lower priority than the `Authority`s that are already present.
```

```rust
// assuming the SchemSystem in built into variable: `schema_system`
// following operation adds new authority with base path `tests` into `schema_system`
schema_system.add_authority(Box::new(FileSystemDocumentAuthority::new(Path::new("test"))));
```

> ### How to create a `SchemsSystem` with given authority or list of authorities?

Reviewer: ```suggestion
### How to create a `SchemaSystem` with given authority or list of authorities?
```

Or even

```suggestion
### Creating an instance of `SchemaSystem`
```

     ]
}
```
The `IslType` can be created as shown below:
```rust
let isl_type = IslType::named(
    // represents the `name` of the defined type
    "my_type_name".to_owned(),
    vec![
        // represents the `type: int` constraint
        IslConstraint::type_constraint(
            IslTypeRef::named("int")
        ),
        // represents `all_of` with anonymous type `{ type: bool }` constraint
        IslConstraint::all_of(
            vec![
                IslTypeRef::anonymous(
                    vec![
                        IslConstraint::type_constraint(
                            IslTypeRef::named("bool")
                        )
                    ]
                )
            ]
        )
    ]
);
```

### How to create `IslSchema`?

```rust
// The `isl-type` defiend in the previous section cna be used here
let isl_schema = IslSchema::new(vec![], vec![isl_type], vec![]);
```

## How to validate an Ion value using a `Schema`?

### Example schema `my_schema.isl`

This file (`my_schema.isl`) defines a new type (`my_int_type`) based on Ion's int type.

```
schema_header::{
  imports: [],
}

type::{
  name: my_int_type,
  type: int,
}

schema_footer::{
}

> ```

Reviewer: This is a repetition of L189-L203. Can you instead have one copy of this sample, and then just refer back to it whenever you need to?
